File: algorithms/sliding_window/239_SlidingWindowMaximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        arr = collections.deque()
        res = [0] * (len(nums) - k + 1)

        for i in range(len(nums)):
            while arr and arr[0] < i - k + 1:
                arr.popleft()
            while arr and nums[i] >= nums[arr[-1]]:
                arr.pop()
            arr.append(i)
            if i >= k - 1:
                res[i - k + 1] = nums[arr[0]]

        return res

File: algorithms/sliding_window/239_SlidingWindowMaximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        if (nums.length == 0) return new int[0];
        Deque<Integer> deq = new LinkedList<Integer>();
        int[] res = new int[nums.length - k + 1];
        for (int i=0; i<k; ++i) {
            while (!deq.isEmpty() && nums[i] >= nums[deq.peekLast()])
                deq.removeLast();
            deq.addLast(i);
        }
        for (int i=k; i<nums.length; ++i) {
            res[i-k] = nums[deq.peekFirst()];
            while (!deq.isEmpty() && deq.peekFirst() <= i-k)
                deq.removeFirst();
            while (!deq.isEmpty() && nums[i] >= nums[deq.peekLast()])
                deq.removeLast();
            deq.addLast(i);
        }
        res[nums.length-k] = nums[deq.peekFirst()];
        return res;
    }
}